Move to sideway/joi

bl4ckb0ne opened the pull request:
Fixes #163 

v16.0.0 breaking changes https://github.com/sideway/joi/issues/2037
v17.0.0 breaking changes https://github.com/sideway/joi/issues/2262

Note on the migration from @hapi/joi to sideway/joi https://github.com/sideway/joi/issues/2411

bl4ckb0ne commented:
@laurence-myers CI fails because of node 8.10, should I bump it or should I fix it?

I am also unable to reproduce locally (even when running on the `circleci/node:8.10` image). The image is also not listed anymore in [circleci supported images][1].

[1]: https://circleci.com/docs/2.0/circleci-images/#nodejs

laurence-myers commented:
Thanks for tackling this! Great work so far. 😁

Yes, let's bump to Node 10. (Node 8 reached EOL in 2019)


On Wed, 2 Dec. 2020, 8:49 am Simon Zeni, <notifications@github.com> wrote:

> @laurence-myers <https://github.com/laurence-myers> CI fails because of
> node 8.10, should I bump it or should I fix it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/joiful-ts/joiful/pull/166#issuecomment-736842187>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABQK4MEVQIXEHU3BGKX6ILTSSVQHXANCNFSM4UJEQ2KA>
> .
>

bl4ckb0ne commented:
Alright looks good on the CI side, I'll let you review it.

codeandcats commented on yarn.lock:
```diff
@@ -290,9 +290,9 @@
 
-"@hapi/joi@15.1.1":
+"joi@15.1.1":
   version "15.1.1"
-  resolved "https://registry.yarnpkg.com/@hapi/joi/-/joi-15.1.1.tgz#c675b8a71296f02833f8d6d243b34c57b8ce19d7"
+  resolved "https://registry.yarnpkg.com/joi/-/joi-15.1.1.tgz#c675b8a71296f02833f8d6d243b34c57b8ce19d7"
```
Hmmm, I'm surprised to see v15 here. You might need to regenerate the `yarn.lock` file.

Try running the following:

```sh
rm -rf ./**/node_modules
yarn
rm ./yarn.lock
yarn
```

bl4ckb0ne replied:
I opened #167 because of the output from it. I'll fix the issue once this patch is merged.

bl4ckb0ne replied:
The commands you gave me updated everything not up to date, do you want that in that patch?

bl4ckb0ne replied:
Nevermind, fixed it with `yarn upgrade joi`

codeandcats commented on test/unit/validation.test.ts:
```diff
@@ -148,7 +158,7 @@ describe('Validation', () => {
             const result = validator.validate(login);
             assertValidateSuccess(result, login);
-            assertValidateInvocation(login);
+            //assertValidateInvocation(login);
```
Instead of removing this check, we should probably check that the schema's `validate` method was called as expected.

bl4ckb0ne replied:
Done!

codeandcats commented on tsconfig.json:
```diff
@@ -16,6 +16,7 @@
         "noUnusedParameters": true,
         "sourceMap": true,
         "strict": true,
+        "strictNullChecks": true,
```
This is covered by the [`strict`](https://www.typescriptlang.org/tsconfig#strict) option.

bl4ckb0ne replied:
Done!

codeandcats commented on test/unit/testUtil.ts:
```diff
@@ -96,7 +96,7 @@ export function assertClassSchemaEquals(
     const schema = getJoiSchema(options.Class, joi);
     const expectedSchema = joi.object().keys(options.expectedSchemaMap);
-    expect(schema).toEqual(expectedSchema);
+    expect(Object.keys(schema!)).toEqual(Object.keys(expectedSchema));
```
Hmm, I'm not sure if checking they both have the same keys is really good enough. Do we know why the previous check now fails? We might be able to simply ignore the specific properties that might be causing the issue?

bl4ckb0ne replied:
Ah yes, this one is a bit tricky. Keeping the original check gives the following error

```
Expected: [object omitted for brievety]
Received: serializes to the same string
```

I also tried to use `toMatchObject` and it gives the following error

```
expect(received).toMatchObject(expected)

    - Expected
    + Received

    @@ -68,11 +68,10 @@
          "untrace": [Function anonymous],
          "valid": [Function anonymous],
          "version": "17.3.0",
          "when": [Function anonymous],
          "x": [Function expression],
    -     Symbol(@hapi/lab/coverage/initialize): [Function trace],
        },
        "$_super": Object {
          "default": [Function bound default],
        },
        "$_temp": Object {
    @@ -157,11 +156,10 @@
                  "untrace": [Function anonymous],
                  "valid": [Function anonymous],
                  "version": "17.3.0",
                  "when": [Function anonymous],
                  "x": [Function expression],
    -             Symbol(@hapi/lab/coverage/initialize): [Function trace],
                },
                "$_super": Object {},
                "$_temp": Object {
                  "ruleset": false,
                  "whens": Object {},
```

Objects are 'almost' equals, but I couldn't find out where this tracing symbol came from. So I went with comparing the keys instead.

codeandcats reviewed (changes requested):
Wow, thank you so much for your contribution @bl4ckb0ne, it's seriously much appreciated. 🏆 🍻 

Unfortunately I don't have much time to review your PR this week. I've had a quick scan over it for now and provided some early feedback, but I'd like to do a more comprehensive review on the weekend when I'll have some free time. In the mean time please have a read over my comments and address/reply at your leisure. 😃

codeandcats commented:
One more thing. Whenever we've changed the API of Joiful we need to include the word `BREAKING CHANGE` in the commit body with a short list of breaking changes. This is because we use commit conventions for knowing how to bump versions. Have a look at the examples [here](https://www.conventionalcommits.org/en/v1.0.0/) for more info.

bl4ckb0ne commented:
@codeandcats do you want me to put the joiful changes in the commit body or a link to the joi release note?

codeandcats commented on src/decorators/any.ts:
```diff
@@ -85,13 +83,13 @@ export interface AnySchemaModifiers {
      */
-    tags(tag: string, ...tags: string[]): this;
-    tags(tags: string | string[]): this;
+    tag(tag: string | string[], ...tags: string[]): this;
+    //tag(tags: string[]): this;
```
Very minor thing but could we revert this change please?

You should be able to call like this:
- `.tag('auth')`
- `.tag('auth', 'identity')`
- `.tag(['auth', 'identity'])`

But not like this:
- `.tag(['auth'], 'identity')`

Indirectly related FYI, I believe @laurence-myers's intention with using `.tag(tag: string, ...tags: string[])` instead of simply `.tag(...tags: string[])` is so at least one tag is required when calling the function with variadic parameters.

bl4ckb0ne replied:
Done!

codeandcats replied:
Looks like this didn't get reverted

codeandcats reviewed (changes requested):
Simon,

I'm so sorry for the late re-review of this fantastic Joiful contribution. Again I want to say we really appreciate your effort!

I've created a PR to your branch that just addresses the weak checking that `assertClassSchemaEquals` function was doing. I think once you've got this in and the other minor thing I mentioned in this review I think we're looking pretty golden.

@laurence-myers I would love it though if you could cast your keen eyes across this too before we merge. Particularly across the new `link` constraint which I admittedly did not spend much time on.

Thanks

bl4ckb0ne commented:
I squashed the commits in 3 distinctive commits (migration to joi 16, migration to joi 17 and the ci image bump). I also added the `BREAKING CHANGES` label with a link to the related release notes.

Let me know if there's anything else I can do.

codeandcats reviewed (changes requested):
Looks good @bl4ckb0ne. Just found that one thing re `tag` that didn't get reverted by the looks of it. 👌

bl4ckb0ne commented:
Woopsy, i just removed it

codeandcats commented on src/decorators/any.ts:
```diff
@@ -85,13 +83,12 @@ export interface AnySchemaModifiers {
      */
-    tags(tag: string, ...tags: string[]): this;
-    tags(tags: string | string[]): this;
+    tag(tag: string | string[], ...tags: string[]): this;
```
This might sound like a nit but I think the original signature with the overload is still the most accurate. Can we please leave it as:

```typescript
    tags(tag: string, ...tags: string[]): this;	    tag(tag: string | string[], ...tags: string[]): this;
    tags(tags: string | string[]): this;
```

or

```typescript
    tags(tag: string, ...tags: string[]): this;	    tag(tag: string | string[], ...tags: string[]): this;
    tags(tags: string[]): this;
```

bl4ckb0ne replied:
My bad, it's fixed.

codeandcats reviewed (changes requested):
Hi @bl4ckb0ne, sorry to be pedantic but can we revert the `tag` change back to its original signature please? After this all looks good on my side. Once this change is in and @laurence-myers is happy, I'm cool with merging this and cutting a new release. 😃

codeandcats reviewed (approved):
Okay this is looking good, there's just one remaining issue I've noticed. The commit containing breaking changes needs to have "BREAKING CHANGE:" in the commit body but the current wording is slightly off so the release process does not bump the package version correctly.

In order to skip another feedback cycle I'm going to merge the PR then amend the commit description in master.

codeandcats commented:
Okay, version joiful v3.0.0 is now published! Please raise an issue if you spot any problems.

Thank you again @bl4ckb0ne for your wonderful contributions and patience. You're a champion. 💯 🏆
